lib/internal/fs/streams.js

```diff
@@ -339,7 +356,17 @@ WriteStream.prototype._write = function(data, encoding, cb) {
     });
   }
 
+  if (this.destroyed) return cb();
```

Comment:
I think this might need to be a `cb(new ERR_STREAM_DESTROYED('write'))`

Reply:
Why? There hasn’t been any error, has there?

Reply:
The write hasn't completed. If we don't send an error here the caller would think the write has completed even though it hasn't.

See, https://github.com/nodejs/node/blob/master/lib/_stream_writable.js#L821

Reply:
I’m worried that introducing an error when there previously was none (or at least not usually) would be semver-major, and I’d prefer to keep this PR as close to just being a fix for the bug as possible.

Reply:
I don't quite see how this could become an issue. It's only if called with a callback (which is very unusual) in which case it's actually a bug if it's not an error.

Though if you are worried about it I guess we can leave it as is. It's an unusual edge case after all. Could we at least have a separate semver-major PR for "correct" behaviour?

Reply:
> Could we at least have a separate semver-major PR for "correct" behaviour?

Yes, that’s what I would prefer here :+1:

Reply:
I share @ronag's concern though: it's tantamount to silently ignoring the write request from the user.

Since this is a correctness issue I don't think it needs to be semver-major?

Reply:
> Since this is a correctness issue I don't think it needs to be semver-major?

Maybe it doesn’t need to be, but I still feel like these are two very different things…

- This PR addresses a race condition that can occur randomly and surprisingly. Fixing it *removes* unexpected errors.
- Adding an error to the callback makes behaviour more consistent, but it would *add* unexpected errors. 

I’d rather not mix the two, and I think we’ve treated other situations where we add errors for consistency as semver-major in the past (and I don’t really see any reason not to do that here, too).

Reply:
I don't think that follows, does it? Right now you'll receive an EBADF error event. While the EBADF error is, er, in error, it at least tells you that the write didn't go through.

(I suppose it could also end up writing to a different file if the fd was reopened in the mean time, which of course is worse than what this PR does.)

Reply:
I guess my reasoning is mostly that this is only relevant when the stream has already been destroyed at this point, and so it should be expected that writes may not finish?

If you feel strongly, I’ll apply @ronag’s suggestion, but I’m still a bit worried about breakage.

Reply:
The way I see it is that you already receive unpredictable EBADF errors now. A predictable error is better, and certainly better than silently dropping data on the floor.

Another way of looking at it: how likely is this change to break existing, functionally correct code? I expect the answer is 'close to zero' - any code that breaks was probably already broken, just not reliably so.

Does that sound reasonable?

Reply:
I’ve pushed a commit with the suggestion … still feeling a bit worried about it but we’ll see if this is problematic